src/server/routes/api/element-association/indra.js

```diff
@@ -0,0 +1,114 @@
+import fetch from 'node-fetch';
+import _ from 'lodash';
+
+import { INDRA_STATEMENTS_URL, INDRA_ENGLISH_ASSEMBLER_URL } from '../../../../config';
+import logger from '../../../logger';
+import { tryPromise } from '../../../../util';
+import { INTERACTION_TYPE } from '../../../../model/element/interaction-type/enum';
+import querystring from 'querystring';
+
+const SUB_MODIFICATION_TYPES = ['Phosphorylation', 'Dephosphorylation', 'Dephosphorylation',
+  'Deubiquitination', 'Methylation', 'Demethylation'];
+const BASE_MODIFICATION_TYPES = ['Sumoylation', 'Desumoylation', 'Hydroxylation', 'Dehydroxylation',
+  'Acetylation', 'Deacetylation', 'Glycolsyation', 'Deglycolsyation', 'Farnesylation',
+  'Defarnesylation', 'Geranylgeranylation', 'Degeranylgeranylation', 'Palmitoylation',
+  'Depalmitoylation', 'Myristoylation', 'Demyristoylation', 'Ribosylation', 'Deribosylation',
+  'Autophosphorylation', 'Transphosphorylation'];
+const TRANSCRIPTION_TRANSLATION_TYPES = ['IncreaseAmount', 'DecreaseAmount'];
+const BINDING_TYPES = ['Complex'];
+
+const getDocuments = pairs => {
+  const getForPair = pair => {
+    let agent0 = pair[0].toUpperCase();
+    let agent1 = pair[1].toUpperCase();
+
+    return tryPromise( () => getInteractions(agent0, agent1) )
+      .then( intns => _.groupBy( intns, 'pmid' ) );
+  };
+
+  const merger = (objValue, srcValue) => {
+    if (_.isArray(objValue)) {
+      return objValue.concat(srcValue);
+    }
+  };
+
+  let promises = pairs.map( getForPair );
+  return Promise.all( promises )
+    .then( res => _.mergeWith( {}, ...res, merger ) );
+};
+
+const transformIntnType = indraType => {
+  if ( SUB_MODIFICATION_TYPES.includes( indraType ) ) {
+    return indraType.toLowerCase();
+  }
+
+  if ( BASE_MODIFICATION_TYPES.includes( indraType ) ) {
+    return INTERACTION_TYPE.MODIFICATION.value;
+  }
+
+  if ( TRANSCRIPTION_TRANSLATION_TYPES.includes( indraType ) ) {
+    return INTERACTION_TYPE.TRANSCRIPTION_TRANSLATION.value;
+  }
+
+  if ( BINDING_TYPES.includes( indraType ) ) {
+    return INTERACTION_TYPE.BINDING.value;
+  }
+
+  return INTERACTION_TYPE.INTERACTION.value;
+};
+
+const getInteractions = (agent0, agent1) => {
```

Comment:
Currently the output format is like:

```
{
   "3291921":[
      {
         "text":"TP53 binds VIM, EGFR, and Oestrogen receptor.",
         "type":"binding",
         "pmid":"3291921"
      }
   ],
   "8887630":[
      {
         "text":"Transcriptionally active TP53 increases the amount of EGFR.",
         "type":"transcription-translation",
         "pmid":"8887630"
      }
   ]
}
```

The keys represents the pmid's. I think it makes sense to switch to the format that you mentioned.

Are you also suggesting to use every evidence rather than using only the first one?

Reply:
@jvwong Is the change in https://github.com/PathwayCommons/factoid/pull/716/commits/d48809131d4aa46c7e0773cbb45e8d21231ce78e enough to resolve this? Or do you also have suggestions for using each evidence of a statement in a way?

Reply:
Fine with me but your `extractInteraction` has to iterate over the `evidence` array.

Reply:
Okay I did that. However I guess that this increased the possibility of having the  interactions with exactly same properties in the same document. I guess they may not be the same interactions actually since they come from different evidences but that different evidence may likely to be pointing the same document. 

I think there are 3 possible ways:

- Keep doing what we are doing now.
- For each indra statment while extracting the interactions of that statement represent all having the same pmid with a single interaction.
- Do that elimination of duplication not only inter the statements but for the whole document. I mean if a document would have some edges with exactly same properties include just one of them (I do not know how likely this is when they are not coming from the same statement though).

Reply:
I see what you mean. As I see it, redundant `element` entries arise from either:
  - Distinct mentions of the same interaction in the paper as detected by a given `source_api` (e.g. reach etc)
  - Different `source_api` reading the same mention of an interaction in a paper

What do you think about just eliminating duplication within each `elements` array. The tradeoff is we throw away this 'signal' (e.g. if a paper mentions the same interaction 100 times it has the same result as 1 mention in another paper.)

Reply:
An alternative is to use the evidence `text` instead of the text assembler output which would remove redundancy associated with #1 above.

Reply:
I updated the code like that.

Reply:
How about keeping both of the assembled text and the evidence text in the output?